src/components/OfflineSearch.ts

```diff
@@ -0,0 +1,173 @@
+        }
+
+        if (!targetNode) {
+            this.setState({ alertMessage: `search offline widget: unable to find grid with the name "${targetName}"` });
+        } else {
+            this.targetWidget = dijitRegistry.byNode(targetNode);
+        }
+    }
+
+    private isValidWidget(): boolean {
+        if (this.targetWidget && this.targetWidget.declaredClass === "mxui.widget.ListView") {
+            if (this.targetWidget._onLoad
+                && this.targetWidget._loadMore
+                && this.targetWidget._renderData
+                && this.targetWidget._datasource
+                && this.targetWidget._datasource.atEnd
+                && typeof this.targetWidget._datasource._pageSize !== "undefined"
+                && (typeof this.targetWidget._datasource._setsize !== "undefined"
+                || typeof this.targetWidget._datasource._setSize !== "undefined")) {
+                    return true;
+            } else {
+                this.setState({ alertMessage: "search offline widget: this Mendix version is incompatible with the offline search widget" });
+            }
+        } else {
+            this.setState({ alertMessage: `search offline widget: supplied target name "${this.props.targetGridName}" is not of the type listview` });
+        }
+
+        return false;
+    }
+
+    private setUpEvents() {
+        this.searchButton.addEventListener("click", this.onClear);
+        this.searchInput.addEventListener("keyup", this.onSearchKeyDown);
+    }
+
+    private setDefaultSearch() {
+        if (this.isValidWidget()) {
+            this.searchInput.value = this.props.defaultQuery;
+            this.updateConstraints(this);
+        }
+    }
+
+    private onSearchKeyDown(event: CustomEvent) {
+        if (this.isValidWidget()) {
+            this.updateButtonVisibility();
+            const searchTimeout = setTimeout(this.updateConstraints(this), 500);
+            clearTimeout(searchTimeout);
+        }
+    }
+
+    private updateButtonVisibility() {
+        if (this.searchInput.value.trim()) {
+            this.setState({ buttonVisibility: "visible" });
+        } else {
+            this.setState({ buttonVisibility: "hidden" });
+        }
+    }
+
+    private updateConstraints(self: OfflineSearch) {
+        const datasource = self.targetWidget._datasource;
+        let constraints = `[${self.props.searchMethod}(${self.props.searchAttribute},'${self.searchInput.value}')]`;
+
+        if (this.props.searchEntity) {
+            constraints = `[${self.props.searchMethod}(${self.props.searchEntity}/${self.props.searchAttribute},'${self.searchInput.value}')]`;
+        }
+        self.searchInput.value.trim() ? datasource._constraints = constraints : datasource._constraints = null;
+        self.targetWidget.update();
+    }
+
+    private onClear(event: CustomEvent) {
+        this.searchInput.value = "";
+        this.updateButtonVisibility();
+        this.updateConstraints(this);
```

Comment:
`(this);` dont thing this is needed.... typescript has better ways of binding this